src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,73 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private List<Product> products;
+    private long maxId;
+    private Object lock = new Object();
+
+    public ArrayListProductDao(){
+        this.products = new ArrayList<>();
+        this.maxId = 0;
+        saveSampleProducts();
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public Product getProduct(Long id) throws NoSuchElementException {
+        return products.stream()
+                .filter(product -> id.equals(product.getId()))
+                .findAny()
+                .get();
```

Comment:
Function prototype contains throws NoSuchElementException, but it's not thrown in fact, in case product is not found null is returned. I'd recommend use orElseThrow() instead of get() with message like "Product with id: " + id + " is not found.", it will throw exception like excepted outside DAO.  Another way is just to return Optional<Product> and no exception.